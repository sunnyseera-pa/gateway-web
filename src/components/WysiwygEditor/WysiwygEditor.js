/** @jsx jsx */
import { cx } from '@emotion/css';
import { jsx } from '@emotion/react';
import { convertToRaw } from 'draft-js';
import { draftToMarkdown } from 'markdown-draft-js';
import PropTypes from 'prop-types';
import React from 'react';
import { Editor } from 'react-draft-wysiwyg';
import LayoutBox from '../LayoutBox';
import { PROP_TYPES_LAYOUTBOX } from '../LayoutBox/LayoutBox.propTypes';
import * as styles from './WysiwygEditor.styles';

import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';

export const WysiwygEditor = ({
    editorState,
    onEditorStateChange,
    onContentStateChange,
    onMarkdownChange,
    toolbar,
    className,
    mt,
    mb,
    ml,
    mr,
    width,
    minWidth,
    maxWidth,
}) => {
    const [markdown, setMarkdown] = React.useState();
    const ref = React.createRef(null);

    const handleOnEditorStateChange = editorStateChanged => {
        console.log('************************* CHANGING');
        const guidanceAsMarkdown = draftToMarkdown(convertToRaw(editorStateChanged.getCurrentContent()));

        onEditorStateChange(editorStateChanged);

        if (markdown !== guidanceAsMarkdown) onMarkdownChange(guidanceAsMarkdown, editorStateChanged);
    };

    React.useEffect(() => {
        setMarkdown(draftToMarkdown(convertToRaw(editorState.getCurrentContent())));
    }, []);

    return (
        <LayoutBox {...{ mt, mb, ml, mr, width, minWidth, maxWidth }} className={cx('ui-WysiwygEditor', className)} css={styles.root()}>
            <Editor
                data-testid='wysiwyg-editor-main'
                wrapperClassName='demo-wrapper'
                editorClassName='wysiwyg-wrapper'
                toolbarClassName='rdw-editor-toolbar'
                editorState={editorState}
                onEditorStateChange={handleOnEditorStateChange}
                onContentStateChange={onContentStateChange}
                ref={ref}
                toolbar={toolbar}
            />
        </LayoutBox>
    );
};

WysiwygEditor.propTypes = {
    onMarkdownChange: PropTypes.func,
    onEditorStateChange: PropTypes.func,
    onContentStateChange: PropTypes.func,
    editorState: PropTypes.shape({
        getCurrentContent: PropTypes.func,
    }).isRequired,
    toolbar: PropTypes.shape({
        options: PropTypes.arrayOf(PropTypes.string),
        inline: {
            inDropdown: PropTypes.bool,
            options: PropTypes.arrayOf(PropTypes.string),
        },
        blockType: {
            inDropdown: PropTypes.bool,
            options: PropTypes.arrayOf(PropTypes.string),
        },
        list: {
            options: PropTypes.arrayOf(PropTypes.string),
        },
        link: {
            options: PropTypes.arrayOf(PropTypes.string),
        },
        history: {
            options: PropTypes.arrayOf(PropTypes.string),
        },
    }),
    ...PROP_TYPES_LAYOUTBOX,
};

WysiwygEditor.defaultProps = {
    onMarkdownChange: () => {},
    onEditorStateChange: () => {},
    onContentStateChange: () => {},
    toolbar: {
        options: ['inline', 'blockType', 'list', 'link', 'history'],
        inline: {
            inDropdown: false,
            options: ['bold', 'italic'],
        },
        blockType: {
            inDropdown: false,
            options: ['Normal', 'H1', 'H2'],
        },
        list: {
            options: ['unordered', 'ordered'],
        },
        link: {
            options: ['link'],
        },
        history: {
            options: ['undo', 'redo'],
        },
    },
};

export default WysiwygEditor;
