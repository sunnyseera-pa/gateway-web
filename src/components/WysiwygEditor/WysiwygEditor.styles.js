import { css } from '@emotion/react';
import { getSize } from '../../configs/theme';

export const root = () => theme => {
    const {
        colors,
        components: {
            Icon: { sizes },
        },
    } = theme;

    return css`
        .wysiwyg-wrapper {
            border: 2px solid ${colors.grey400};
            border-top: none;
            padding: 0px ${getSize(4, theme)};
            min-height: 400px;
            max-height: 520px;
        }

        .rdw-editor-toolbar {
            padding: 6px 5px 0;
            border-radius: 2px;
            border: 1px solid ${colors.grey100};

            border-top: solid 2px ${colors.grey400} !important;
            border-left: solid 2px ${colors.grey400} !important;
            border-right: solid 2px ${colors.grey400} !important;

            display: flex;
            justify-content: flex-start;
            background: white;
            flex-wrap: wrap;
            font-size: 15px;
            margin-bottom: 0px !important;
            user-select: none;
        }

        .wysiwyg-main-card {
            margin-bottom: 0px;
            border-bottom: solid 1px $background;
        }
    `;
};
