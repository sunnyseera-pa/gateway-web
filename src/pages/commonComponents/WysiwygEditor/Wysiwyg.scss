@import '../../../css/variables';

.wysiwyg-wrapper {
	border: solid 2px $gray-400;
	border-top: none;
	padding: 0px 16px;
	min-height: 400px;
	max-height: 520px;
}

.rdw-editor-toolbar {
	padding: 6px 5px 0;
	border-radius: 2px;
	border: 1px solid #f1f1f1;

	border-top: solid 2px $gray-400 !important;
	border-left: solid 2px $gray-400 !important;
	border-right: solid 2px $gray-400 !important;

	display: flex;
	justify-content: flex-start;
	background: white;
	flex-wrap: wrap;
	font-size: 15px;
	margin-bottom: 0px !important;
	user-select: none;
}

.wysiwyg-main-card {
	margin-bottom: 0px;
	border-bottom: solid 1px $background;
}
