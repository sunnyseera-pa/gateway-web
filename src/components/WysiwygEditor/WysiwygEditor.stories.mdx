import { Meta, Story, Canvas } from '@storybook/addon-docs';
import { convertFromRaw, EditorState } from 'draft-js';
import { markdownToDraft } from 'markdown-draft-js';

import WysiwygEditor from './WysiwygEditor';

<Meta title='Core/WysiwygEditor' component={WysiwygEditor} />

<Canvas>
    <Story name='Default'>
        {props => {
            const contentState = convertFromRaw(markdownToDraft('Some sample markup'));
            return <WysiwygEditor editorState={EditorState.createWithContent(contentState)} />;
        }}
    </Story>
</Canvas>
