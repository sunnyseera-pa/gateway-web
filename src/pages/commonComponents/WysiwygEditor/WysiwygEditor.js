import { convertToRaw } from 'draft-js';
import { draftToMarkdown } from 'markdown-draft-js';
import React from 'react';
import { Editor } from 'react-draft-wysiwyg';
import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';
import './Wysiwyg.scss';

export const WysiwygEditor = ({ editorState, onEditorStateChange, onContentStateChange, onMarkdownChange }) => {
    const [markdown, setMarkdown] = React.useState();
    const ref = React.createRef(null);

    const handleOnEditorStateChange = editorStateChanged => {
        const guidanceAsMarkdown = draftToMarkdown(convertToRaw(editorStateChanged.getCurrentContent()));

        onEditorStateChange(editorStateChanged);

        if (markdown !== guidanceAsMarkdown) onMarkdownChange(guidanceAsMarkdown, editorStateChanged);
    };


    React.useEffect(() => {
        setMarkdown(draftToMarkdown(convertToRaw(editorState.getCurrentContent())));
    }, []);

    return (
        <Editor
            data-testid='wysiwyg-editor-main'
            wrapperClassName='demo-wrapper'
            editorClassName='wysiwyg-wrapper'
            toolbarClassName='rdw-editor-toolbar'
            editorState={editorState}
            onEditorStateChange={handleOnEditorStateChange}
            onContentStateChange={onContentStateChange}
            ref={ref}

            toolbar={{
                options: ['inline', 'blockType', 'list', 'link', 'history'],
                inline: {
                    inDropdown: false,
                    options: ['bold', 'italic'],
                },
                blockType: {
                    inDropdown: false,
                    options: ['Normal', 'H1', 'H2'],
                },
                list: {
                    options: ['unordered', 'ordered'],
                },
                link: {
                    options: ['link'],
                },
                history: {
                    options: ['undo', 'redo'],
                },
            }}
        />
    );
};

WysiwygEditor.defaultProps = {
    onMarkdownChange: () => {},
    onEditorStateChange: () => {},
    onContentStateChange: () => {},

};

export default WysiwygEditor;
