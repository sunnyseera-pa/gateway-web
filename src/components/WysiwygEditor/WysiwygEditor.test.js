import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react';

import { createEvent } from '@testing-library/dom';
import { convertFromRaw, EditorState } from 'draft-js';
import { markdownToDraft } from 'markdown-draft-js';

import WysiwygEditor from '.';

jest.mock('react-draft-wysiwyg', () => ({
    ...jest.requireActual('react-draft-wysiwyg'),
    Editor: ({ onEditorStateChange }) => {
        return (
            <input
                onChange={x => {
                    console.log('X', x.target);
                    onEditorStateChange(x);
                }}
                data-testid='editor'
            />
        );
    },
}));

const contentState = convertFromRaw(markdownToDraft('Initial content'));
const editorState = EditorState.createWithContent(contentState);

const props = {
    editorState,
    onEditorStateChange: jest.fn(),
    onContentStateChange: jest.fn(),
    onMarkdownChange: jest.fn(),
};

let wrapper;

describe('Given the WysiwygEditor component', () => {
    describe('When it is rendered', () => {
        beforeAll(() => {
            wrapper = render(<WysiwygEditor {...props} />, {
                wrapper: Providers,
            });
        });

        describe('And the content is changed', () => {
            beforeAll(() => {
                const editor = wrapper.getByTestId('editor');

                fireEvent(editor, createEvent('change', editor, editorState), { EventType: 'CustomEvent' });
            });
            it('Then calls the correct methods', async () => {
                expect(wrapper.container).toMatchSnapshot();

                await waitFor(() => {
                    expect(props.onEditorStateChange).toHaveBeenCalled();
                });
            });
        });
    });
});
