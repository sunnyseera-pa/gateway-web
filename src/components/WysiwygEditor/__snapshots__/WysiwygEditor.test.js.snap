// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the WysiwygEditor component When it is rendered And the content is changed Then calls the correct methods 1`] = `
<div>
  <div
    class="ui-LayoutBox ui-WysiwygEditor css-gkdke2-WysiwygEditor css-1k6rz1e"
  >
    <input
      data-testid="editor"
    />
  </div>
</div>
`;
